Big blocks

zander opened the pull request:
This reverts BIP109. Classic will no longer vote for BIP109.

This reuses the code and starts voting for a new bit which is a protocol
upgrade and the only change in that one is that the maximum block size is
removed.

This new fork is activated using bit 26, while requiring signalling for
BIP9 making the block version 0x28000000 When 75% of 1000 blocks have this
bit set we lock in this change and have a 4 week wait period.
After that period blocks no longer have a size limit.

Miners can continue to use the -blockmaxsize argument or bitcoin.conf
setting to limit the size of blocks they generate!  All nodes can also use
the `-blocksizeacceptlimit` option to limit the size of the blocks accepted
from other miners.

This change simplifies some of the designs that were introduced for BIP9, like the overloading of the  IsSuperMajority method and the txdb getter/setter is now made substantially simpler too. Overdesigns are good to remove.

Last, I moved the fork-bit to be namespaced inside CBlockHeader, being a tiny little bit more structured instead of everything in the global namespace.

zander committed:
Update code to reflect the new protocol-upgrade

This reverts BIP109. Classic will no longer vote for BIP109.

This reuses the code and starts voting for a new bit which is a protocol
upgrade and the only change in that one is that the maximum block size is
removed.

This new fork is activated using bit 26, while requiring signalling for
BIP9 making the block version 0x28000000 When 85% of 1000 blocks have this
bit set we lock in this change and have a 4 week wait period.
After that period blocks no longer have a size limit.

Miners can continue to use the -blockmaxsize argument or bitcoin.conf
setting to limit the size of blocks they generate!  All nodes can also use
the `-blocksizeacceptlimit` option to limit the size of the blocks accepted
from other miners.

zander committed:
Choose another random bit

zander committed:
Introduce new config option blocksizeacceptlimit

This option allows a user to configure the behaviour of the
block-acceptance algorithm. A node will reject incoming blocks that are
larger than the block size accept limit and as such a miner would be able
to decide to limit blocks it will build on top of based on size.

The blocksizeacceptlimit is a number in 100KB increments.
The default is 20, which makes the default accepted block 2MB.

gmaxwell commented:
Why bother with the version bit when you already accept oversized blocks without it? It appear to have no effect on the software's consensus behavior.

zander commented:
@gmaxwell The pull request doesn't accept oversized blocks without it.

gmaxwell commented:
"To be able to be 'ready' for a spontaneous fork Bitcoin Classic removes the check for max block size even if the upgrade has not been triggered."

dskloet commented:
Why increase the trigger from 750 blocks to 850 blocks?

dskloet commented on src/init.cpp:
```diff
@@ -489,6 +489,8 @@ std::string HelpMessage(HelpMessageMode mode)
     strUsage += HelpMessageOpt("-bytespersigop", strprintf(_("Minimum bytes per sigop in transactions we relay and mine (default: %u)"), DEFAULT_BYTES_PER_SIGOP));
     strUsage += HelpMessageOpt("-datacarrier", strprintf(_("Relay and mine data carrier transactions (default: %u)"), DEFAULT_ACCEPT_DATACARRIER));
     strUsage += HelpMessageOpt("-datacarriersize", strprintf(_("Maximum size of data in data carrier transactions we relay and mine (default: %u)"), MAX_OP_RETURN_RELAY));
+    // called excessiveblocksize in BU
```
Why abbreviate Bitcoin Unlimited? It might be less clear in some future.

dskloet commented on src/init.cpp:
```diff
@@ -489,6 +489,8 @@ std::string HelpMessage(HelpMessageMode mode)
     strUsage += HelpMessageOpt("-datacarrier", strprintf(_("Relay and mine data carrier transactions (default: %u)"), DEFAULT_ACCEPT_DATACARRIER));
     strUsage += HelpMessageOpt("-datacarriersize", strprintf(_("Maximum size of data in data carrier transactions we relay and mine (default: %u)"), MAX_OP_RETURN_RELAY));
+    // called excessiveblocksize in BU
+    strUsage += HelpMessageOpt("-blocksizeacceptlimit=<n>", strprintf("This node will not accept blocks of this size or larger. Unit is multiples of 100KB (default: %u)", DEFAULT_BLOCK_ACCEPT_SIZE));
```
Looking at the code, blocks of this size are accepted, but not larger, which is what you'd expect from a "limit".
For the doc string?

zander replied:
I'll change it to;

> This node will not accept blocks larger than this limit.

dskloet commented on src/main.cpp:
```diff
@@ -2046,18 +2040,26 @@ static int64_t nTimeTotal = 0;
 
 static bool DidBlockTriggerSizeFork(const CBlock &block, const CBlockIndex *pindex, const CChainParams &chainparams)
```
Rename to DidBlockTriggerSizeRemovalFork?

dskloet commented on src/main.cpp:
```diff
@@ -2046,18 +2040,26 @@ static int64_t nTimeTotal = 0;
+    for (int i = 0; i < params.nMajorityWindow && current; ++i)
+    {
+        if ((current->nVersion & CBlockHeader::SIZE_FORK_BIT) == CBlockHeader::SIZE_FORK_BIT) {
+            if (++found > params.nActivateSizeForkMajority)
```
Use && instead of nested if?

greatwolf replied:
I thought the same at first but that `++found` is a little tricky. Does the C++ standard say anything about how sequence points apply to pre-increment operators? Does the short circuit ensure `++` does not happen?

dskloet replied:
Yes, from [the standard](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n4296.pdf):

> Unlike &, && guarantees left-to-right
> evaluation: the second operand is not evaluated if the first operand is false.

zander replied:
C specifies the order of parsing and if one is 'false' then the rest will not be done.
see man 7 operator for order.

In this case I could have joined them. I guess I should ;)

dskloet commented on src/miner.cpp:
```diff
@@ -113,6 +114,22 @@ CBlockTemplate* Mining::CreateNewBlock(const CChainParams& chainparams) const
+    uint32_t nBlockMaxSize = std::max<uint32_t>(1000, GetArg("-blockmaxsize", DEFAULT_BLOCK_MAX_SIZE));
+    if (pblock->nTime < sizeForkTime.load()) { // before the protocol upgrade, limit size.
+        nBlockMaxSize = std::min<uint32_t>(nBlockMaxSize, MAX_BLOCK_SIZE);
+    }
```
I prefer curly brackets but for consistency should this if have them removed like all the others?

zander commented:
@dskloet
The numbers in this pull request are not final, they are suggestions and this PR is one way to go, out of many possible ones. The number one complaint we got in BIP109 was the low acceptance, lets try something larger.

dskloet commented on src/policy/policy.h:
```diff
@@ -27,6 +27,9 @@ static const unsigned int MAX_P2SH_SIGOPS = 15;
 static const unsigned int MAX_STANDARD_TX_SIGOPS = MAX_BLOCK_SIGOPS/5;
 /** Default for -maxmempool, maximum megabytes of mempool memory usage */
 static const unsigned int DEFAULT_MAX_MEMPOOL_SIZE = 300;
+/** Default for -blocksizeacceptlimit */
```
Comment that it's multiples of 100kB. Actually, why is it multiples of 100kB?

zander replied:
Thanks, will fix.
Its multiples of 100kB because its much easier to write 20 than 2000000. I do think people want to allow things like 1.5MB which would make the finest grained change 100kB.

dskloet replied:
It does seem odd to specify 2MB as 20.
How about making it multiples of 1MB but allowing floating point numbers so 1.5MB can be specified as 1.5?

zander replied:
The argument processing in Bitcoin doesn't support that, so I'd have to do some more work ;)  May be worth it, I don't know.

greatwolf replied:
Would you guys be open to using User-Defined literals? This is a post C++11 feature though so not sure if it's a problem building on platforms.

solarj replied:
This is a sensitive option, if you set it below 10 then you will create a permanent chain split (a tightening of the rules is a soft fork, and with minor hash power you can extend the minority chain, with major hash power you orphan others

zander replied:
@greatwolf I'm not sure if user-defined literals can help us parsing user-provided strings. Please note that Classic already requires C++11.

dskloet commented:
Can you say something about how this interacts with BU?

dskloet commented:
My impression of the 75% complainers was that they prefer 95%, which tells me they just prefer it doesn't trigger at all. Have people asked for 85% or is this the kind of compromise that makes nobody happy?

zander commented:
> Can you say something about how this interacts with BU?

The change is compatible with BU and a discussion starter on safe upgrade paths.
I explicitly added support for the commandline option from BU for the limits.

> My impression of the 75% complainers was that they prefer 95%, which tells me they just prefer it doesn't trigger at all.

You may have a point. I think the 95% stems from a lack of understanding of how Bitcoin works. We may have to work on that and just keep the voting to 75%

dskloet commented:
> The change is compatible with BU

What does that mean? If 850/1000 blocks are mined by BU will Classic trigger? Will/does Classic signal acceptance of larger blocks? Before the trigger? After the trigger?

greatwolf commented:
So can we assume this initial PR is for making inroads into making classic more inline with BU's concept of emergent consensus? For example, are there plans to add other parameters from BU like acceptance depth? What about modifying user-agent of classic to include `ADx:EBx` like BU to communicate what's being accepted by nodes running classic?

zander commented:
@dskloet 
The phrase "The change is compatible with BU" means that this change would make BU and Classic behave the same after a hard fork. There is no activation concept in BU currently, so naturally the details are not compatible. Yet. The entire point is to find a way to do the forking safely and in a coordinated manner.

This change makes Classic have only configurable limits on both blocks it creates and receives. This is basically according to our roadmap and planned for Q3. So it was overdue ;)

dskloet commented:
My questions still stand :)

> If 850/1000 blocks are mined by BU will Classic trigger? Will/does Classic signal acceptance of larger blocks? Before the trigger? After the trigger?

I didn't ask if BU would activate. I asked if BU signaling could cause Classic to activate.

gmaxwell commented:
Why have my questions been deleted?

zander commented:
@greatwolf I'm warming up to the idea of using the ADx:EBx etc. I won't have time to do it this month, so if I would welcome a pull request on that matter.

zander committed:
Take care of all the feedback items.

The biggest difference is that the parameter 'limit' is now a floating
point number parsed in megabytes. So you could make the limit to be
 "1.6"

gmaxwell commented:
Why have my questions been deleted? -- then why was my message asking about the deletion (with its three upvotes) deleted?

In the very first response to this PR, I asked what purpose the version signaling accomplished, since this patch performs the hardforking change without coordination by accepting oversized blocks:
Zander responded:

>    @gmaxwell The pull request doesn't accept oversized blocks without it.

To which I replied, quoting from the pull request text: "Bitcoin Classic removes the check for max block size even if the upgrade has not been triggered" and from the commit request, "The default is 20, which makes the default accepted block 2MB."

All these comments have now been removed.  Then a comment asking about it (same 95% same text as this message) seems to have gone missing too.

zander commented:
I pushed an update that addresses all comments.

dgenr8 commented on src/main.cpp:
```diff
@@ -3128,14 +3125,23 @@ bool CheckBlock(const CBlock& block, CValidationState& state, bool fCheckPOW, bo
+        }
     }
+    if (nSizeLimit <= 0)
+        nSizeLimit = DEFAULT_BLOCK_ACCEPT_SIZE * 1E5;
```
1E6?

zander replied:
There is more wrong with that line, thanks for pointing it out.

dskloet replied:
Why wasn't this caught by a test?

dgenr8 commented:
It would be very helpul if you were to write a precise text description of the parameters, their defaults, and their interaction with the proposed version bit.  Without this, it's difficult to know which behavior is intentional.

I suggest not using any floating-point for block size calculations.

gmaxwell commented:
Why have my questions been deleted? -- then why was my message asking about the deletion (with its three upvotes) deleted?

In the very first response to this PR, I asked what purpose the version signaling accomplished, since this patch performs the hardforking change without coordination by accepting oversized blocks:
Zander responded:

> @gmaxwell The pull request doesn't accept oversized blocks without it.

To which I replied, quoting from the pull request text: "Bitcoin Classic removes the check for max block size even if the upgrade has not been triggered" and from the commit request, "The default is 20, which makes the default accepted block 2MB."

All these comments have now been removed. Then a comment asking about it (same 95% same text as this message) seems to have gone missing too.

zander commented:
@gmaxwell Your initial question was answered, you didn't like the answer and didn't bother reading the code to see why you were wrong in the first place, and you then start to repost your question that was already answered, again and again. You are not being productive, this is called trolling. This is not the first time and you are wasting everyones time.

This is not acceptable behaviour and I will point out that if you want to continue working with Classic you will have to stop trolling and actually contribute to the conversation.

gmaxwell commented:
@zander Your response directly contradicted the text of the PR.  I asked for a clarification and your response was to delete my comment and take no further action?!  I had assumed someone just hit the wrong button...

greatwolf commented on src/main.cpp:
```diff
@@ -3128,14 +3125,23 @@ bool CheckBlock(const CBlock& block, CValidationState& state, bool fCheckPOW, bo
+    std::int32_t nSizeLimit = GetArg("-excessiveblocksize",  -1) * 1E6;
+    auto userlimit = mapArgs.find("-blocksizeacceptlimit");
+    if (userlimit != mapArgs.end()) {
+        double limitInMB = atof(userlimit->second.c_str());
```
How about `std::stod(userlimit->second);` instead?

greatwolf commented on src/main.cpp:
```diff
@@ -2050,12 +2050,9 @@ static bool DidBlockTriggerSizeFork(const CBlock &block, const CBlockIndex *pind
-                return true;
-        }
+    for (int i = 0; i < params.nMajorityWindow && current; ++i) {
+        if ((current->nVersion & CBlockHeader::SIZE_FORK_BIT) == CBlockHeader::SIZE_FORK_BIT && ++found >= params.nActivateSizeForkMajority)
```
isn't `== CBlockHeader::SIZE_FORK_BIT` kind of redundant?

gmaxwell commented:
It should be noted, for the record, that Thomas Zander is removing any comments which are remotely critical of the technology in this pull request.  As a result the discussion here misrepresents the level and the results of the review that it has received.

rebroad commented:
Yeah, @gmaxwell, you're in Zanderland now and your pertinent questions (aka
trolling) will not be tolerated here!
On 27 Oct 2016 16:45, "TomZ" <notifications@github.com> wrote:

> @gmaxwell <https://github.com/gmaxwell> Your initial question was
> answered, you didn't like the answer and didn't bother reading the code to
> see why you were wrong in the first place, and you then start to repost
> your question that was already answered, again and again. You are not being
> productive, this is called trolling. This is not the first time and you are
> wasting everyones time.
>
> This is not acceptable behaviour and I will point out that if you want to
> continue working with Classic you will have to stop trolling and actually
> contribute to the conversation.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoinclassic/bitcoinclassic/pull/191#issuecomment-256583279>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdZq-EndG_Uwzj-XsL-80ZylF_H6vMkks5q4GTBgaJpZM4Kf8Ev>
> .
>

zander committed:
Don't use floats to determine block size.

greatwolf commented on src/main.cpp:
```diff
@@ -3136,7 +3136,7 @@ bool CheckBlock(const CBlock& block, CValidationState& state, bool fCheckPOW, bo
     }
     if (nSizeLimit <= 0)
-        nSizeLimit = DEFAULT_BLOCK_ACCEPT_SIZE * 1E5;
+        nSizeLimit = static_cast<int32_t>(DEFAULT_BLOCK_ACCEPT_SIZE * 10) * 1E5;
```
Should probably leave a comment to indicate that the decimal truncation is intentional. The commit message could be confusing because you're still doing a float operation at the end -- `1E5` is a float literal.

For consistency maybe `std::int32_t` or change `std::uint32_t` to just `uint32_t`.

zander commented:
Closing as #195 supersedes this proposal.
